### DISCUSSION TITLE
why context canceled?

### DISCUSSION THREAD

**Discussion by gonta1026** [urql-graphql/urql#3465] [Q&A] [answered]
I'm using the <Link /> component to transition screens, and when I make a request with the same query, the client makes so many requests that I get a context canceld on the golang server.
This phenomenon is rare when the requestPolicy is set to network-only.
If we set requestPolicy to cache-first, which is the default, this problem does not seem to occur, but due to the nature of the application we are developing, we would like to avoid using cache as much as possible.

However, due to the nature of the application we are developing, we would like to avoid using cache as much as possible.
- Why this problem is occurring?
- Why is a large number of queries being requested just by moving from one screen to another?
- How to deal with it

### sample code

```tsx
// a.tsx
export default function APage() {
  const [{ data }] = useQuery({
    query: GetUserDocument,
    requestPolicy: 'network-only',
  })
  console.log('APage data', { data })
  return (
    <div style={{ width: '200px', margin: '20px auto' }}>
      <Link href={'/b'}>to B</Link>
    </div>
  )
}

```

```tsx
// b.tsx
export default function BPage() {
  const [{ data }] = useQuery({
    query: GetUserDocument,
    requestPolicy: 'network-only',
  })
  console.log('BPage data', { data })
  return (
    <div style={{ width: '200px', margin: '20px auto' }}>
      <Link href={'/a'}>to A</Link>
    </div>
  )
}

```

```
// server log
{
  "origin": "http://localhost:3003/",
  "error": "context canceled",
  "errorVerbose": "context canceled\n(1) attached stack trace\n  -- stack trace:
  | ...
Wraps: (2) context canceled
"errCode": "INTERNAL_SERVER_ERROR"
}
```

### screen shot
It seems that a large number of queries are issued by the client after just one screen transition.
All of these queries deal with GetUserDocument.

<img width="767" alt="スクリーンショット 2023-12-13 19 15 22" src="https://github.com/urql-graphql/urql/assets/45672476/d24e9f11-279c-45f5-840d-049c6b1f21b3">

### version
- "@urql/core": "^4.0.10"
- "next": "14.0.4",

**Comment by JoviDeCroock** [comment]
Could you make a reproduction or at least give us something to work off, your details are a bit low 😅 it could be that you are using `Suspense` and clearing your client every time by not catching it in a boundary located underneath your `Provider`.

**Reply by gonta1026** [reply]
> Could you make a reproduction or at least give us something to work off, your details are a bit low 😅 it could be that you are using Suspense and clearing your client every time by not catching it in a boundary located underneath your Provider.

Thank you for your comment. I am not sure I fully understand, but it seems to be the following, although the files involved are few.
I don't think there is anything in particular that I am doing,,,,

```tsx
// provider
import { Provider } from 'urql'
import { urqlClient } from '@/lib'

export const URQLProvider = ({ children }: { children: React.ReactNode }) => {
  return <Provider value={urqlClient}>{children}</Provider>
}
```

```
// _app.tsx
import { AppProps } from 'next/app'
import React from 'react'
import { URQLProvider } from '@/providers/urqlProvider'

export default function MyApp({ Component, pageProps }: AppProps) {
  return (
    <URQLProvider>
      <Component pageProps={pageProps} />
    </URQLProvider>
  )
}

```

**Reply by JoviDeCroock** [reply]
That still doesn't show me anything but my assumption is that _app rerenders and you keep creating a new urql client 😅

**Reply by gonta1026** [reply]
> That still doesn't show me anything

I forgot to attach the urqlClient file!💦
```
// urqlClient
import { Client, cacheExchange, fetchExchange } from '@urql/core'
import { devtoolsExchange } from '@urql/devtools'
import fetch from 'cross-fetch'
import { env } from '@/config'

const newUrqlClient = ({ url }: { url: string }) => {
  return new Client({
    fetch,
    url,
    exchanges: [cacheExchange, devtoolsExchange, fetchExchange],
    requestPolicy: 'network-only',
    fetchOptions: {
      headers: {
        Accept: '*/*',
        'Content-Type': 'application/json',
      },
    },
  })
}

export const urqlClient = newUrqlClient({ url: env.NEXT_PUBLIC_GRAPHQL_PROXY_ENDPOINT })

```

> but my assumption is that _app rerenders and you keep creating a new urql client 😅

I tried to plant a log, but I couldn't find any new urql clients being new.

**Reply by gonta1026** [reply]
If you have any problems here, I would appreciate your comments.

https://github.com/urql-graphql/urql/discussions/3465#discussioncomment-7883699

**Reply by gonta1026** [reply]
@JoviDeCroock 

> _app rerenders and you keep creating a new urql client 😅

I am aware that it is not unusual for _app.tsx to re-render, but I see this as a problem,

- A duplicate request is sent.
- context cancel occurs

I would like to know if there is a solution to the two problems. If not it may be difficult to adopt urql in the project I am participating in.


[error-sample-repository](https://github.com/gonta1026/urql-sample)

**Reply by JoviDeCroock** [reply, marked as answer]
## Urql side

I have looked at your reproduction and honestly, I don't know why the first request goes through, this check is due to the `cache-check` which tries to see synchronously whether or not there is a cached item available.

- [cache check](https://github.com/urql-graphql/urql/blob/main/packages/react-urql/src/hooks/useQuery.ts#L337-L351) 
- [fetchExchange safety](https://github.com/urql-graphql/urql/blob/main/packages/core/src/internal/fetchSource.ts#L177-L179)

What this seems to signal is that the `effect` cleanup phase of React seems to be delayed longer than a microtick. When I change this to wait a tick i.e. `setTimeout(x, 0)` then it correctly cancels the request.

There are a few solutions I see here from the urql side of things

- Change the cache-check to use a `cache-only` policy
- Change to waiting for a tick

The issue here being that we don't know whether this react change will be long-term, Next uses an experimental version of React for everything....

CC @kitten 

EDIT: after discussion with my co-maintainer we decided not to act on this, this might be related to strict-mode behaving weirdly in next,... it stands out that Next uses experimental react and us as library maintainers aren't being updated on the latest developments there.
EDIT2: in your reproduction you can prevent this issue by hoisting `Layout` to `_app` as it creates multiple versions of `Layout` while you do not leverage a cache meaning all of them need to fetch their own data

## Your side

When I look at your reproduction I see that you are using `getServerSideProps`, however the cache produced by that isn't being used in your application as [documented here](https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#legacy-nextjs-pages) this used `next-urql` but you can restore this as well with [similar logic](https://github.com/urql-graphql/urql/blob/next-urql%405.0.2/packages/next-urql/src/with-urql-client.ts#L83) if you don't want to use that integration.

The idea here being that you fetch your data on the server but never hydrate it in your client-side cache, just wanted to highlight this in case you need to do this in the future 😅

**Reply by gonta1026** [reply]
Thanks for checking.

> Change the cache-check to use a cache-only policy

The application I am developing is currently implemented without using cache as much as possible for safety reasons, so it is currently difficult to support cache-only.

> Change to waiting for a tick

 I don't understand the code and would appreciate it if someone could fix this...!
 
> Your side

Thank you. I understand

 ---
 
However, in my application it is mandatory to use `network only` to get new information from the DB each time in a basic api request and I hope urql will solve this problem.

**Reply by gonta1026** [reply]
> EDIT: after discussion with my co-maintainer we decided not to act on this, this might be related to strict-mode behaving weirdly in next,... it stands out that Next uses experimental react and us as library maintainers aren't being updated on the latest developments there.

I understand the situation...! Thank you for your consideration.

> EDIT2: in your reproduction you can prevent this issue by hoisting Layout to _app as it creates multiple versions of Layout while you do not leverage a cache meaning all of them need to fetch their own data

I can confirm that this can also be solved by putting it in _app, but I am assuming that different Layout components will be used for each screen, and putting this Layout in _app is difficult in my application.

**Comment by kitten** [comment]
We'd need more information here.
There's templates for reproductions at: https://github.com/urql-graphql/urql/tree/main/examples

However, I went ahead and created two sandboxes to quickly validate this:
- `app/` folder Next.js (i.e. app router): https://stackblitz.com/edit/github-qdzf6z
- `pages/` folder Next.js: https://stackblitz.com/edit/github-qdzf6z-xcoiud

Those are derived from our examples, which are made for reproductions, basically.

To summarise the problem here, there's some gotchas when changing the request policy for every operation to `cache-and-network`. And there's even more gotchas when changing it to `network-only`. Notes here: https://github.com/urql-graphql/urql/blob/73628b71a023a9b0be5b4e280c48c7ebf9a977fc/packages/core/src/client.ts#L153-L156

This is **not** a problem though! It should still work 😅 
However, I think, it may lead to an issue in React, in strict mode specifically, in some cases. But that's hard for me to tell without a reproduction. 🤔

**Reply by gonta1026** [reply]
Thanks for the detailed explanation, it helped me understand better.
I would like to investigate further by looking at the documentation, sample code, and using @urql/exchange-request-policy🤗

**Reply by gonta1026** [reply]
I have a question about @urql/exchange-request-policy that caused a type error.

https://github.com/urql-graphql/urql/discussions/3466

**Reply by gonta1026** [reply]
If you have any problems here, I would appreciate your comments.

https://github.com/urql-graphql/urql/discussions/3465#discussioncomment-7883699

**Comment by gonta1026** [comment]
I'm still adjusting the code and doing some research,
- Getting Hoge with getServerSideProps.
- Get Fuga in CSR.
- context canceled 😢

but When I did urql with CSR only, without SSR, the problem did not cause the context cancel issue.

**Comment by gonta1026** [comment]
I have tried to minimize the code here to make sure it works, but on rare occasions I get an error when I make a screen transition.
I would appreciate your comments if you could look at the code and see if there are any problems.


https://github.com/gonta1026/urql-sample

<img width="1410" alt="スクリーンショット 2023-12-18 18 05 46" src="https://github.com/urql-graphql/urql/assets/45672476/6c81ec22-82e8-423d-85a6-4833782e6ccd">

**Reply by gonta1026** [reply]
For some reason, the same query appears to be requested twice on the same page and one of the requests is cancelled.
<img width="1432" alt="スクリーンショット 2023-12-18 18 40 21" src="https://github.com/urql-graphql/urql/assets/45672476/0ebbf70a-3636-44f9-ab30-7599e3395e78">
